libass/ass_fontselect.c

```diff
@@ -166,6 +170,111 @@ static ASS_FontProviderFuncs ft_funcs = {
     .destroy_font      = destroy_font_ft,
 };
 
+#if defined(_WIN32) && !defined(__CYGWIN__)
+/*
+ * Output parameters are only valid if the call was sucessful, ignore them otherwise.
+ * \param path          input; path of the dir to open, may be UTF-8 or the codepage used by fopen
+ * \param file_context  output
+ * \param find_handle   output; needs to be closed
+ * \param out_length    output
+ * \return utf-16le version of path; needs to be freed
+ */
+static wchar_t *win32_open_dir(const char *path, WIN32_FIND_DATAW *file_context, HANDLE *find_handle, size_t *out_length)
+{
+    UINT cur_cp = CP_UTF8;
+    size_t tmp;
+
+    if (!path || !(tmp = strlen(path)) || tmp > SIZE_MAX - 3)
+        return NULL;
+
+    char* valid_search_dir = malloc(tmp + 3);
+    if (!valid_search_dir)
+        return NULL;
+    strcpy(valid_search_dir, path);
+    switch (valid_search_dir[tmp - 1]) {
+    case '\\':
+    case '/':
+        break;
+    default:
+        valid_search_dir[tmp++] = '\\';
+    }
+    valid_search_dir[tmp++] = '*';
+    valid_search_dir[tmp] = '\0';
+
+    size_t wlen;
+    wchar_t *wpath = ass_convert_to_utf16le(valid_search_dir, cur_cp, &wlen);
+    if (!wpath || (*find_handle = FindFirstFileExW(wpath, 0, file_context, 0, NULL, 0)) == INVALID_HANDLE_VALUE) {
+        free(wpath);
+        cur_cp = CP_ACP;
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+        if (!AreFileApisANSI()) cur_cp = CP_OEMCP;
+#endif
+        wpath = ass_convert_to_utf16le(valid_search_dir, cur_cp, &wlen);
+        if (wpath) {
+            if ((*find_handle = FindFirstFileExW(wpath, 0, file_context, 0, NULL, 0)) == INVALID_HANDLE_VALUE) {
+                free(wpath);
+                wpath = NULL;
+                goto clean;
+            }
+        }
+        else
+            goto clean;
```

Comment:
Style Nit: Imho the `else` is better placed at the same line as the closing `}` from the `if`-block